### PR TITLE
Clean up and refactor code around Multiple Input Sets

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Refactor out the  boilerplate code to rearrange the multiple input sets between:
[[inp0, inp0', inp0''], inp1] (the argument form allowed in the API)
and
[(inp0, inp1), (inp0', inp1), (inp0'', inp1)]  (list of argument packs suitable for single operator)

Refactor deprecation handling and constant promotion to dedicated functions.

Fix the code flow in the `_preprocess_inputs`, so all constants are resolved there, before
multiple input sets are processed - this allows to reuse the created constant operators.
Previously it was done in two separate places, where constant op for data like numpy arrays
were created in `_preprocess_inputs` and ScalarConstants were replaced in OperatorInstance
constructor.

The innermost conditions in the main loop of `_preprocess_inputs` were always satisfied,
this should be now fixed.

Docstring and comments were added.

### Key points relevant for the review:
Ops, multiple input sets.
Please check if there 

### Tests:

- [x] Existing tests apply
This is refactoring, whole DALI should work without change.
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [x] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
I'm trying to push one more commit here, but to no avail. :(

File: dali/python/nvidia/dali/ops/__init__.py [hunk 1]
```diff
@@ -70,7 +84,7 @@ def _separate_kwargs(kwargs, arg_input_type=_DataNode):
     def is_arg_input_type(x):
         return isinstance(x, arg_input_type)
 
-    def is_call_arg(name, value):
+    def is_arg_input(name, value):
```
Comment (mzient) [inline dali/python/nvidia/dali/ops/__init__.py:87]: This is not a correct name. The argument "name" is passed to `__call__` even though it's not an argument input. The question is whether we can somehow rectify this and pass `name` to the `__init__` instead.

**Comment by klecki** [comment]
!build

**Review by stiepan — APPROVED** [review]

File: dali/python/nvidia/dali/ops/_operator_utils.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    """Repack list from [[a, b, c], [a', b', c'], ....]
+    to [fn(a, a', ...), fn(b, b', ...), fn(c, c', ...)]
+    where fn can be `tuple` or `list`
+    Assume that all elements of input have the same length
+    """
+    output_list = []
+    arg_list_len = len(sets[0])
+    for i in range(arg_list_len):
+        output_list.append(fn(input_set[i] for input_set in sets))
+    return output_list
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/ops/_operator_utils.py:26]: It's an old code, we can probably rewrite it to something either shorter or nicer :P

It's used only after `_unify_lists` was called on the inputs or when we got outputs from several instances of operators. In both cases we are guaranteed that the lengths of the inner elements are equal.

File: dali/python/nvidia/dali/ops/__init__.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops/__init__.py]: Done

**Review by mzient — APPROVED** [review]

**Pull request merged**